internal/verification/authenticate_stats.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package verification
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/google/exposure-notifications-server/internal/verification/database"
+	"github.com/google/exposure-notifications-server/internal/verification/model"
+	"github.com/google/exposure-notifications-server/pkg/logging"
+)
+
+// AuthenticateStatsToken parse the provided JWT and determines if it is an authorized stats request.
+func (v *Verifier) AuthenticateStatsToken(ctx context.Context, rawToken string) (int64, error) {
+	logger := logging.FromContext(ctx)
+
+	var healthAuthorityID int64
+	var claims *jwt.StandardClaims
+
+	token, err := jwt.ParseWithClaims(rawToken, &jwt.StandardClaims{}, func(token *jwt.Token) (interface{}, error) {
+		if method, ok := token.Method.(*jwt.SigningMethodECDSA); !ok || method.Name != jwt.SigningMethodES256.Name {
+			return nil, fmt.Errorf("unsupported signing method, must be %v", jwt.SigningMethodES256.Name)
+		}
+
+		kidHeader, ok := token.Header["kid"]
+		if !ok {
+			err := errors.New("missing 'kid' header in token")
```

Comment:
These should all probably just return an error instead of logging. Move the logger higher